### PR TITLE
blog for tempest plugin entrypoint macro

### PR DESCRIPTION
**Pull request by chkumar246**

### CODE REVIEW

**Comment by chkumar246** [comment]
@apevec @hguemar @javierpena Please have a look.

File: source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+---
+title: Writing rpm-macros for OpenStack
+author: chandankumar
+date: 2017-01-11 14:00:00 IST
+tags: blogs,openstack,rdo,rpm-macros
+---
+
+rpm-macros is a short string, always prefixed by % and generally surrounded by curly brackets ({}) which RPM will convert to a different and usually longer string.
+Some macros can take arguments and some can be quite complex.
+In RHEL, Centos and Fedora, Macros are provided by rpm package and from [redhat-rpm-config](https://apps.fedoraproject.org/packages/redhat-rpm-config).
```
Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: CentOS is the official spelling

Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: Macros -> macros

File: source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
+---
+title: Writing rpm-macros for OpenStack
+author: chandankumar
+date: 2017-01-11 14:00:00 IST
+tags: blogs,openstack,rdo,rpm-macros
+---
+
+rpm-macros is a short string, always prefixed by % and generally surrounded by curly brackets ({}) which RPM will convert to a different and usually longer string.
```
Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: rpm-macros -> RPM macro

File: source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+Some macros can take arguments and some can be quite complex.
+In RHEL, Centos and Fedora, Macros are provided by rpm package and from [redhat-rpm-config](https://apps.fedoraproject.org/packages/redhat-rpm-config).
+In RDO, openstack macros are provided by [openstack-macros](http://cbs.centos.org/koji/packageinfo?packageID=3792) which comes from upstream [rpm-packaging](http://git.openstack.org/cgit/openstack/rpm-packaging/) project.
```
Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: openstack -> OpenStack

Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: Maybe link directly to http://git.openstack.org/cgit/openstack/rpm-packaging/tree/openstack/openstack-macros ? Not sure.

File: source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+To evaluate a rpm macro:
+
+    $ rpm --eval %{_bindir}
+
+%py_build is the commonly used rpm-macro in RDO OpenStack packages which points to *python setup.py build* process.
+
+    $ rpm --eval %py_build
+
+**Motivation behind writing rpm-macro for OpenStack**
```
Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: writing rpm-macro for OpenStack -> writing a new RPM macro for OpenStack packages

File: source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+Currently, Tempest provides a external test plugin interface which enables anyone to integrate an external test suite as a part of tempest run and each service tempest plugin has a entrypoint
```
Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: a external -> an external

Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: tempest -> Tempest

Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: a entrypoint -> an entrypoint

File: source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+defined in setup.cfg through which tempest discovers it and list the tempest plugins.
```
Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: tempest -> Tempest

File: source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+For example:
+
+    tempest.test_plugins =
+        heat_tests = heat_integrationtests.plugin:HeatTempestPlugin
+
+In RDO OpenStack services Packages, In tree tempest plugins packages are provided by openstack-{service}-tests subpackage but the tempest plugin entrypoint is provided by the main package openstack-%{service}.
```
Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: Packages -> RPM packages

Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: In tree -> in-tree

Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: tempest -> Tempest

File: source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+So once you have a working OpenStack environment with tempest installed having no test subpackage installed. Then we tried to run tempest commands you would have encountered "No module heat_integrationtests.plugin found"
+and you endup installing a hell lot of packages to fix this. The basic reason for the above error is tempest plugin entry point is installed by main openstack package but files pointing to entrypoint are not found.
```
Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: endup -> end up
s/hell//
tempest -> Tempest
openstack -> OpenStack

File: source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+To fix the above issue we have decided to seperate out the tempest plugin entrypoint from main package and move it to openstack-{service}-tests subpackage during rpmbuild process by creating a fake tempest plugin entry point
+for all RDO services packages. Since it is a massive and similar change affecting all openstack services packages.
+So i have created %py2_entrypoint macro which is available in OpenStack Ocata release.
```
Comment (apevec) [inline source/blog/2017-01-11-writing-rpm-macros-for-openstack.html.md]: s/. So i/, I/

**Comment by amoralej** [comment]
Nice post Chandan!!

**Pull request merged**